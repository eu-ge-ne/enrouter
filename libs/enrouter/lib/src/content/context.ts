import { createContext, useContext } from "react";

import type { Content } from "./mod.js";

const ContentContext = createContext<Content>({
  match: {
    route: {
      path: "",
      test: {
        keys: [],
        pattern: new RegExp(""),
      },
      modules: [],
      loaded: false,
      elements: {},
    },
    location: "",
    isFull: false,
    params: {},
  },
});

export const ContentProvider = ContentContext.Provider;

export function useContent(): Content {
  return useContext(ContentContext);
}
