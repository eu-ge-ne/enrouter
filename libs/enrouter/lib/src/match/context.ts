import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
