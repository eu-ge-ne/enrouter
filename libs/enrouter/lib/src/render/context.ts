import { createContext } from "react";

import type { RouteNodes } from "./mod.js";

export const RouteRenderContext = createContext<RouteNodes>({
  match: {
    handler: {
      route: {
        path: "",
        mod: [],
        link: [[], []],
      },
      test: { keys: [], pattern: new RegExp("") },
      modules: [],
    },
    location: "",
    isFull: false,
    params: {},
  },
});
