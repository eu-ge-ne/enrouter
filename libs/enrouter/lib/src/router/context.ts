import { createContext, useContext } from "react";

import type { Route } from "#lib/route/mod.js";

export interface TRouterContext {
  routes: Route;
  location: string;
  navigate: (to: string) => void;
  ctx: unknown;
}

const RouterContext = createContext<TRouterContext>({
  routes: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },
  location: "",
  navigate: () => undefined,
  ctx: undefined,
});

export const RouterProvider = RouterContext.Provider;

export function useRouter(): TRouterContext {
  return useContext(RouterContext);
}
