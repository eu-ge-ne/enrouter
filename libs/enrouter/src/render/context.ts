import { createContext } from "react";

import type { RouteNodes } from "./mod.js";

export const RouteRenderContext = createContext<RouteNodes>({
  match: {
    params: {},
    handler: {
      route: {
        path: "",
        mod: [],
        link: { css: [], mod: [] },
      },
      test: { keys: [], pattern: new RegExp("") },
      modules: [],
    },
  },
});
