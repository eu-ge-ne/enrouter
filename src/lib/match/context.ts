import { createContext, useContext } from "react";

import type { Match } from "./match.js";

const MatchesContext = createContext<Match[]>([]);
const MatchIndexContext = createContext<number>(-1);

export const MatchesProvider = MatchesContext.Provider;
export const MatchIndexProvider = MatchIndexContext.Provider;

export interface MatchContext {
  matches: Match[];
  matchIndex: number;
  match: Match | undefined;
  firstMatch: Match | undefined;
  nextMatch: Match | undefined;
  lastMatch: Match | undefined;
  voidMatch: Match | undefined;
  isExactMatch: boolean;
}

export function useMatchContext(): MatchContext {
  const matches = useContext(MatchesContext);
  const matchIndex = useContext(MatchIndexContext);

  const match = matches[matchIndex];
  const firstMatch = matches[0];
  const nextMatch = matches[matchIndex + 1];
  const lastMatch = matches.at(-1);
  const voidMatch = matches.findLast((x) => x.route?.elements._void);

  const isExactMatch = Boolean(lastMatch?.route);

  return {
    matches,
    matchIndex,
    match,
    firstMatch,
    nextMatch,
    lastMatch,
    voidMatch,
    isExactMatch,
  };
}

export function useMatch(path: string): Match | undefined {
  return useContext(MatchesContext).find((x) => x.route?.path === path);
}
