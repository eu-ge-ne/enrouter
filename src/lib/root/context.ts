import { type ComponentType, createContext, useContext } from "react";

export interface RootParams {
  root?: ComponentType;
  fallback?: Record<string, ComponentType>;
}

const RootParamsContext = createContext<RootParams>({});

export const RootParamsProvider = RootParamsContext.Provider;

export function useRootParams(): RootParams {
  return useContext(RootParamsContext);
}
