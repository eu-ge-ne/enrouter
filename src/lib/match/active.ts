import { useMatches } from "./context.js";

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({
  path,
  loose,
  value: [yes, no],
}: UseActiveParams<T>): T {
  const matches = useMatches();

  const match = matches.find((x) => x.route?.path === path);

  if (!match?.route) {
    return no;
  }

  if (loose) {
    return yes;
  }

  if (match !== matches.at(-1)) {
    return no;
  }

  return match?.route ? yes : no;
}
