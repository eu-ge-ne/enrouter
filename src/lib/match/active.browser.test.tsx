import type { FC, PropsWithChildren } from "react";
import { describe, test, expect } from "vitest";
import { render } from "vitest-browser-react";
import * as regexparam from "regexparam";

import type { Match } from "./match.js";
import { MatchesProvider } from "./context.js";
import { useActive } from "./active.js";

const testId = "test-wrapper";

const wrapper: FC<PropsWithChildren> = ({ children }) => (
  <div data-testid={testId}>{children}</div>
);

describe("match", () => {
  describe("useActive", () => {
    test("2 matches strict", async () => {
      const matches: Match[] = [
        {
          route: {
            path: "/",
            test: regexparam.parse("/", true),
            modules: [],
            loaded: true,
            elements: {},
          },
          location: "/",
          params: {},
        },
        {
          route: {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {},
          },
          location: "/abc",
          params: {},
        },
      ];

      const Test: FC = () => {
        const match1 = useActive({
          path: "/",
          value: ["/ is active", "/ is not active"],
        });
        const match2 = useActive({
          path: "/abc",
          value: ["/abc is active", "/abc is not active"],
        });
        const match3 = useActive({
          path: "/xyz",
          value: ["/xyz is active", "/xyz is not active"],
        });

        return (
          <div>
            <p>{match1}</p>
            <p>{match2}</p>
            <p>{match3}</p>
          </div>
        );
      };

      const screen = render(
        <MatchesProvider value={matches}>
          <Test />
        </MatchesProvider>,
        { wrapper },
      );

      await expect.element(screen.getByTestId(testId)).toBeVisible();

      expect(screen.container).toMatchSnapshot();
    });

    test("2 matches loose", async () => {
      const matches: Match[] = [
        {
          route: {
            path: "/",
            test: regexparam.parse("/", true),
            modules: [],
            loaded: true,
            elements: {},
          },
          location: "/",
          params: {},
        },
        {
          route: {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {},
          },
          location: "/abc",
          params: {},
        },
      ];

      const Test: FC = () => {
        const match1 = useActive({
          path: "/",
          loose: true,
          value: ["/ is active", "/ is not active"],
        });
        const match2 = useActive({
          path: "/abc",
          loose: true,
          value: ["/abc is active", "/abc is not active"],
        });
        const match3 = useActive({
          path: "/xyz",
          loose: true,
          value: ["/xyz is active", "/xyz is not active"],
        });

        return (
          <div>
            <p>{match1}</p>
            <p>{match2}</p>
            <p>{match3}</p>
          </div>
        );
      };

      const screen = render(
        <MatchesProvider value={matches}>
          <Test />
        </MatchesProvider>,
        { wrapper },
      );

      await expect.element(screen.getByTestId(testId)).toBeVisible();

      expect(screen.container).toMatchSnapshot();
    });
  });
});
