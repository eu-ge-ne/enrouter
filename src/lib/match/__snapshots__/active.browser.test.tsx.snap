// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`match > useActive > 2 matches loose 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      <p>
        / is active
      </p>
      <p>
        /abc is active
      </p>
      <p>
        /xyz is not active
      </p>
    </div>
  </div>
</div>
`;

exports[`match > useActive > 2 matches strict 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      <p>
        / is not active
      </p>
      <p>
        /abc is active
      </p>
      <p>
        /xyz is not active
      </p>
    </div>
  </div>
</div>
`;
